File: Android/app/src/main/java/com/example/android/canvas/Rectangle.kt
package com.example.android.canvas

import android.content.Context
import android.view.View
import com.example.android.SocketHandler
import com.example.android.client.ClientInfo
import org.apache.batik.anim.dom.SVGOMRectElement
import org.apache.batik.dom.AbstractDocument
import org.w3c.dom.Element
import java.lang.Float.min
import kotlin.math.abs

open class Rectangle(private var drawingId: Int?,
                     prefix: String, owner: AbstractDocument):
                SVGOMRectElement(prefix, owner), Tool {
    override var currentX = 0f
    override var currentY = 0f
    override var selected = false
    override var str = "<rect "
    override var startTransformPoint = Point(0f, 0f)
    override var totalTranslation = Point(0f,0f)
    override var totalScaling = Point(0f,0f)
    override var scalingPositions = HashMap<Point, Point>()
    private var selectionOffset = 0f
    //override var drawingID = drawingId
    override var contentID: Int?= null
    //var abstractTool = AbstractTool(this)

    override fun touchStart(eventX: Float, eventY: Float, svgRoot: Element) {
        this.setAttribute("x", eventX.toString())
        this.setAttribute("y", eventY.toString())
        this.setAttribute("width", "0")
        this.setAttribute("height", "0")
        this.setAttribute("transform", "translate(0,0)")
        this.setAttribute("stroke-width", "${DrawingUtils.thickness}")
        this.setAttribute("stroke", DrawingUtils.primaryColor)
        this.setAttribute("fill", DrawingUtils.secondaryColor)
        //svgRoot.appendChild(this)
        requestCreation()
    }

    override fun touchMove(context: Context, eventX: Float, eventY: Float) {
        val width = abs(eventX - this.getAttribute("x").toFloat())
        val height = abs(eventY - this.getAttribute("y").toFloat())
        this.setAttribute("width", width.toString())
        this.setAttribute("height", height.toString())
        currentY = eventY
        currentX = eventX
        val x = this.getAttribute("x").toFloat()
        val y = this.getAttribute("y").toFloat()

        this.setAttribute("y", min(currentY,y).toString())
        this.setAttribute("x", min(currentX,x).toString())
        if(contentID != null){
            sendProgressToServer(DrawingStatus.InProgress)
        }
    }

    override fun setCriticalValues(){
        val xCert = this.getAttribute("x").toFloat()
        val yCert = this.getAttribute("y").toFloat()
        val widthCert = this.getAttribute("width").toFloat()
        val heightCert = this.getAttribute("height").toFloat()
        startTransformPoint.x = xCert + (widthCert/2)
        startTransformPoint.y = yCert + (heightCert/2)
    }

    override fun touchUp() {
        setCriticalValues()
        calculateScalingPositions()
        select()
    }

    override fun getString(): String {
        str = ""
        try{
            str += getOriginalString()
            if(selected){
                getSelectionString()
                getScalingPositionsString()
            }
        }catch(e: Exception){}

        return str
    }

    override fun getOriginalString(): String {
        var result = "<rect "
        val x = this.getAttribute("x").toFloat()
        val y = this.getAttribute("y").toFloat()
        val width = this.getAttribute("width")
        val height = this.getAttribute("height")
        val transform = this.getAttribute("transform")
        val stroke = this.getAttribute("stroke")
        val strokeWidth = this.getAttribute("stroke-width")
        val fill = this.getAttribute("fill")

        result += "x=\"${x}\" "
        result += "y=\"${y}\" "
        result += "width=\"$width\" "
        result += "height=\"$height\" "
        result += "transform=\"$transform\""
        result += " stroke=\"$stroke\""
        result += " stroke-width=\"$strokeWidth\""
        result += " fill=\"$fill\""

        result += "/>\n"
        return result
    }

    override fun inTranslationZone(eventX: Float, eventY: Float): Boolean{
        val x = this.getAttribute("x").toFloat()
        val y = this.getAttribute("y").toFloat()
        val width = this.getAttribute("width").toFloat()
        val height = this.getAttribute("height").toFloat()

        val isInXAxes = eventX <= x + width + (2*selectionOffset) + totalTranslation.x
            && eventX >= x - selectionOffset + totalTranslation.x
        val isInYAxes = eventY <= y + height + (2*selectionOffset)+ totalTranslation.y
            && eventY >= y - selectionOffset + totalTranslation.y
        return isInXAxes && isInYAxes
    }

    override fun scale(view: View, scalePoint: Point , direction: Point) {
        val x = this.getAttribute("x").toFloat()
        val y = this.getAttribute("y").toFloat()
        val width = this.getAttribute("width").toFloat()
        val height = this.getAttribute("height").toFloat()
        val minPoint = Point(x , y)
        val maxPoint = Point(x + width, y + height)
        if(direction.x == -1f){
            println(minPoint.x)
            minPoint.x += scalePoint.x
            println(minPoint.x)
            currentX = minPoint.x
        }
        else if(direction.x == 1f){
            maxPoint.x += scalePoint.x
        }
        if(direction.y == -1f){
            minPoint.y += scalePoint.y
            currentY = minPoint.y
        }
        else if(direction.y == 1f){
            maxPoint.y += scalePoint.y
        }

        if(minPoint.x >= maxPoint.x){
            direction.x *= -1
        }

        if(minPoint.y >= maxPoint.y){
            direction.y *= -1
        }
        this.setAttribute("y", min(minPoint.y,maxPoint.y).toString())
        this.setAttribute("x", min(minPoint.x,maxPoint.x).toString())
        this.setAttribute("width", abs(maxPoint.x - minPoint.x).toString())
        this.setAttribute("height", abs(maxPoint.y - minPoint.y).toString())
        sendProgressToServer(DrawingStatus.Selected)
    }

    override fun translate(view: View, translationPoint: Point) {
        totalTranslation.makeEqualTo(translationPoint)
        this.setAttribute("transform",
            "translate(${totalTranslation.x},${totalTranslation.y})")
        sendProgressToServer(DrawingStatus.Selected)

    }

    override fun getSelectionString() {
        str += "<rect "
        val x = this.getAttribute("x").toFloat()
        val y = this.getAttribute("y").toFloat()
        val width = this.getAttribute("width").toFloat()
        val height = this.getAttribute("height").toFloat()
        val transform = this.getAttribute("transform")

        str += "x=\"${x}\" "
        str += "y=\"${y}\" "
        str += "width=\"${width}\" "

        str += "height=\"${height}\" "
        str += "transform=\"${transform}\""

        str += " stroke=\"#0000FF\""
        str += " stroke-width=\"3\""
        str += " fill=\"none\""
        str += " stroke-dasharray=\"4\""
        str += "/>\n"
    }

    override fun calculateScalingPositions() {
        scalingPositions.clear()
        val width = this.getAttribute("width").toFloat()
        val height = this.getAttribute("height").toFloat()
        val x = this.getAttribute("x").toFloat()
        val y = this.getAttribute("y").toFloat()
        val firstPos = Point(x + totalTranslation.x, y + totalTranslation.y)
        val firstDirection = Point(-1f, -1f)
        scalingPositions[firstPos] = firstDirection

        val secondPos = Point(x + (width/2) + totalTranslation.x, y + totalTranslation.y)
        scalingPositions[secondPos] = Point(0f,-1f)

        val thirdPos = Point(x + width + totalTranslation.x, y + totalTranslation.y)
        scalingPositions[thirdPos] = Point(1f, -1f)

        val forthPos = Point(x + width + totalTranslation.x,
            y + (height/2) + totalTranslation.y)
        scalingPositions[forthPos] = Point(1f, 0f)

        val fifthPos = Point(x + width + totalTranslation.x,
            y + height + totalTranslation.y)
        scalingPositions[fifthPos] = Point(1f, 1f)

        val sixthPos = Point(x + (width/2) + totalTranslation.x
            , y + height + totalTranslation.y)
        scalingPositions[sixthPos] = Point(0f, 1f)

        val seventhPos = Point(x + totalTranslation.x
            , y + height + totalTranslation.y)
        scalingPositions[seventhPos] = Point(-1f, 1f)

        val eighthPos = Point(x + totalTranslation.x ,
            y + (height/2) + totalTranslation.y)
        scalingPositions[eighthPos] = Point(-1f, 0f)
    }

    override fun getScalingPoint(point: Point): MutableMap.MutableEntry<Point, Point>?{
        for(item in scalingPositions){
            val position = item.key
            val x = position.x - radius
            val y = position.y - radius
            val width = (position.x + radius) - x
            val height = (position.y + radius) - y
            val inXAxes = point.x >= x && point.x <= x+ width
            val inYAxes = point.y >= y && point.y <= y + height
            if(inXAxes && inYAxes){
                return item
            }
        }
        return null
    }

    override fun getScalingPositionsString(){
        calculateScalingPositions()
        for(item in scalingPositions){
            val position = item.key
            val x = position.x - radius
            val y = position.y - radius
            val width = (position.x + radius) - x
            val height = (position.y + radius) - y
            str += "<rect x=\"$x\" y=\"$y\" width=\"$width\"" +
                " height=\"$height\" stroke=\"#CBCB28\" fill=\"#CBCB28\"/>\n"
        }
    }

    override fun parse(parceableString: String?){
        val xRegex = Regex("""x="([-?0-9.?]*)"""")
        val matchX = xRegex.find(parceableString!!, 1)
        this.setAttribute("x", matchX!!.groups[1]!!.value)
        val yRegex = Regex("""y="([-?0-9.?]*)"""")
        val matchY = yRegex.find(parceableString,1)
        this.setAttribute("y", matchY!!.groups[1]!!.value)
        val widthRegex = Regex("""width="([-?0-9.?]*)""")
        val matchWidth = widthRegex.find(parceableString,1)
        this.setAttribute("width", matchWidth!!.groups[1]!!.value)
        val heightRegex = Regex("""height="([-?0-9.?]*)"""")
        val matchHeight = heightRegex.find(parceableString,1)
        this.setAttribute("height", matchHeight!!.groups[1]!!.value)

        //Commune between tools
        val translateRegex = Regex("""translate\(([-?0-9.?]+),([-?0-9.?]+)\)""")
        val matchTranslate = translateRegex.find(parceableString, 1)
        totalTranslation.x = matchTranslate!!.groups[1]!!.value.toFloat()
        totalTranslation.y = matchTranslate.groups[2]!!.value.toFloat()
        this.setAttribute("transform",
            "translate(${totalTranslation.x},${totalTranslation.y})")
        //strokeParse
        val strokeRegex = Regex("""stroke="([#a-zA-Z0-9]+)"""")
        val matchStroke = strokeRegex.find(parceableString, 1)
        this.setAttribute("stroke", matchStroke!!.groups[1]!!.value)
        val strokeWidthRegex = Regex("""stroke-width="([0-9]+)"""")
        val matchStrokeWidth = strokeWidthRegex.find(parceableString, 1)
        this.setAttribute("stroke-width", matchStrokeWidth!!.groups[1]!!.value)
        val fillRegex = Regex("""fill="([#a-zA-Z0-9]+ |none)"""")
        val matchFill = fillRegex.find(parceableString)
        this.setAttribute("fill", matchFill!!.groups[1]!!.value)
        setCriticalValues()
    }

    protected fun sendProgressToServer(status: DrawingStatus){
        val drawingContent = ContentDrawingSocket(
            drawingId = drawingId, userId = ClientInfo.userId,
            contentId = contentID, drawing= getOriginalString(),
            status = status, toolName = rectString)
        val socket = SocketHandler.getDrawingSocket()
        socket.emit("drawingToServer", drawingContent.toJson())
    }

    protected fun requestCreation(){
        SocketHandler.getDrawingSocket()
            .emit("createDrawingContent",
                RequestCreation(drawingId).toJson())
    }

    override  fun unselect(){
        sendProgressToServer(DrawingStatus.Done)
        selected = false
    }

    override fun select(){
        selected = true
        sendProgressToServer(DrawingStatus.Selected)
    }

    override fun delete(){
        sendProgressToServer(DrawingStatus.Deleted)
    }

    override fun updateThickness() {
        this.setAttribute("stroke-width", "${DrawingUtils.thickness}")
        sendProgressToServer(DrawingStatus.Selected)
    }

    override fun updatePrimaryColor() {
        this.setAttribute("stroke", DrawingUtils.primaryColor)
        sendProgressToServer(DrawingStatus.Selected)
    }

    override fun updateSecondaryColor() {
        this.setAttribute("fill", DrawingUtils.secondaryColor)
        sendProgressToServer(DrawingStatus.Selected)
    }
}

File: angular-client/src/app/services/drawing-tools/rectangle.ts
// import { DrawingStatus } from '@models/DrawingMeta';
import { ElementRef, Renderer2 } from '@angular/core';
import { ColorPickingService } from '@services/color-picker/color-picking.service';
import { InteractionService } from '@services/interaction/interaction.service';
import { ChosenColors } from '@src/app/models/ChosenColors';
import { DrawingContent, DrawingStatus } from '@src/app/models/DrawingMeta';
import { userColorMap } from '../drawing/drawing.service';
//import { DrawingContent, DrawingStatus } from '@src/app/models/DrawingMeta';
import { SocketService } from '../socket/socket.service';
import { DrawingTool } from './drawing-tool';
//import { ActiveDrawing, UserToken } from '../static-services/user_token';
// import { DrawingTool } from './drawing-tool';
import { Point } from './point';
//import { Shape } from './shape';
import { RECT_TOOL_NAME } from './tool-names';
import { ShapeTypes, ToolsAttributes } from './tools-attributes';

const DEFPRIM = '#000000ff'
const DEFSEC =  '#ffffffff'
const RADUIS = 10;
const X_REGEX = new RegExp(`x="([-?0-9.?]*)"`);
const Y_REGEX = new RegExp(`y="([-?0-9.?]*)"`);
const WIDTH_REGEX = new RegExp(`width="([-?0-9.?]*)"`);
const HEIGHT_REGEX = new RegExp(`height="([-?0-9.?]*)"`);
const FILL_REGEX = new RegExp(`fill="([#a-zA-Z0-9]+| none)"`);
const STROKE_WIDTH_REGEX = new RegExp(`stroke-width="([0-9]+)"`)
const STROKE_REGEX = new RegExp(`stroke="([#a-zA-Z0-9]+)"`)
const TRANSLATE_REGEX = new RegExp(
  /translate\((-?\d+(?:\.\d*)?),(-?\d+(?:\.\d*)?)\)/
)
export class Rectangle implements  DrawingTool{
  attr!:ToolsAttributes;
  toolName = RECT_TOOL_NAME;
  currentX: number = 0;
  currentY: number = 0;
  str: string= "<rect ";
  selected: boolean = false;
  startTransformPoint: Point = new Point(0,0);
  totalTranslation: Point = new Point(0,0);
  totalScaling: Point = new Point(0,0);
  scalingPositions: Map<Point, Point> = new Map<Point,Point>();
  contentId!: number;
  userId: string;
  element: SVGRectElement;
  drawingId!: number;
  //mouseIsDown: boolean = false;
  primaryColor!: string;
  secondaryColor!: string;
  //pointsArray: Point[];
  selectionOffset: number = 0;
  constructor(
    private interactionService: InteractionService,
    private colorPick: ColorPickingService,
    private socketService: SocketService,
    userId: string,
    private renderer: Renderer2,
    private canvas: ElementRef,
  ) {
    this.element = this.renderer.createElement('rect', 'svg') as SVGRectElement;
    //this.attr= {shapeLineThickness: DEF_LINE_THICKNESS, shapeType: ShapeTypes.BOTH};
    let attr = this.interactionService.toolsAttributes$.value;
    this.attr = {shapeLineThickness: attr.shapeLineThickness, shapeType: attr.shapeType}
    this.userId = userId;
    this.primaryColor = colorPick.cData.primaryColor;
    this.secondaryColor = colorPick.cData.secondaryColor;
  }

  onMouseDown(event: MouseEvent): void {
    //this.mouseIsDown = true;
    let position = Point.rpositionMouse(event, this.canvas.nativeElement);
    //this.pointsArray.push(position)
    this.renderer.setAttribute(this.element, "x", position.x.toString())
    this.renderer.setAttribute(this.element, "y", position.y.toString())
    this.renderer.setAttribute(this.element, "width", "0");
    this.renderer.setAttribute(this.element, "height", "0")
    this.renderer.setAttribute(this.element, "transform", "translate(0,0)");
    //this.renderer.setAttribute(this.element, "stroke", this.primaryColor);
    this.renderer.setAttribute(this.element, "stroke-width", this.attr.shapeLineThickness!.toString())
    //this.renderer.setAttribute(this.element, "fill", this.secondaryColor);
    if(this.attr.shapeType === ShapeTypes.OUTLINE){
      this.renderer.setAttribute(this.element, "stroke", this.primaryColor);
      this.renderer.setAttribute(this.element, "fill", "none");  
    }
    else if(this.attr.shapeType === ShapeTypes.FULL){
      this.renderer.setAttribute(this.element, "stroke", "none");
      this.renderer.setAttribute(this.element, "fill", this.secondaryColor);
    }
    else{
      this.renderer.setAttribute(this.element, "stroke", this.primaryColor);
      this.renderer.setAttribute(this.element, "fill", this.secondaryColor);
    }
    this.requestCreation()
  }

  requestCreation(){
    this.socketService.createDrawingContentRequest({drawingId: this.drawingId});
  }
  onMouseUp(e: MouseEvent): void {
    let x = parseFloat(this.element.getAttribute('x')!)
    let y = parseFloat(this.element.getAttribute('y')!)
    this.renderer.setAttribute(this.element, 'x', `${x + this.totalTranslation.x}`)
    this.renderer.setAttribute(this.element, 'y', `${y + this.totalTranslation.y}`)
    this.totalTranslation.x = 0;
    this.totalTranslation.y = 0;
    this.renderer.setAttribute(this.element, "transform", "translate(0,0)");
    this.setCriticalValues();
    this.calculateScalingPositions();
    this.select();
    //this.mouseIsDown = false;
    //throw new Error('Method not implemented.');
  }
  onMouseMove(event: MouseEvent): void {
    //throw new Error('Method not implemented.');
    let position = Point.rpositionMouse(event, this.canvas.nativeElement)
    let width = Math.abs(position.x - parseFloat(this.element.getAttribute('x')!))
    let height = Math.abs(position.y - parseFloat(this.element.getAttribute('y')!));
    this.renderer.setAttribute(this.element, 'width', width.toString())
    this.renderer.setAttribute(this.element, "height", height.toString())
    this.currentY = position.y;
    this.currentX = position.x;
    let x = parseFloat(this.element.getAttribute('x')!)
    let y = parseFloat(this.element.getAttribute('y')!)
    this.renderer.setAttribute(this.element,'y', Math.min(this.currentY,y).toString())
    this.renderer.setAttribute(this.element, 'x', Math.min(this.currentX, x).toString())
    if(this.contentId!== undefined && this.contentId!== null){
      this.sendProgressToServer(DrawingStatus.InProgress);
    }
  }
  sendProgressToServer(drawingStatus: DrawingStatus){
    let drawingContent: DrawingContent = {drawingId: this.drawingId, userId: this.userId,
      id: this.contentId, content: this.getOriginalString(), status: drawingStatus, toolName: this.toolName}
    this.socketService.sendDrawingToServer(drawingContent);
  }
  getString(): string {
    //throw new Error('Method not implemented.');
    this.str = "";
    this.str += this.getOriginalString();
    if(this.selected){
      this.getSelectionString();
      this.getScalingPositionsString();
    }
    return this.str;
  }
  getOriginalString(): string {
    //throw new Error('Method not implemented.');
    let result = "<rect "
    let x = parseFloat(this.element.getAttribute('x')!)
    let y = parseFloat(this.element.getAttribute('y')!)
    let width = this.element.getAttribute('width');
    let height = this.element.getAttribute('height');
    let transform = this.element.getAttribute('transform');
    let stroke = this.element.getAttribute('stroke');
    let strokeWidth = this.element.getAttribute('stroke-width');
    let fill = this.element.getAttribute('fill');

    result += `x="${x}" `;
    result += `y="${y}" `;
    result += `width="${width}" `;
    result += `height="${height}" `;
    result += `transform="${transform}" `
    result += `stroke="${stroke}" `;
    result += `stroke-width="${strokeWidth}" `;
    result += `fill="${fill}"`;
    result += "/>\n";
    return result;
  }
  inTranslationZone(event: MouseEvent): boolean {
    //throw new Error('Method not implemented.');
    let position = Point.rpositionMouse(event, this.canvas.nativeElement);
    let x= parseFloat(this.element.getAttribute('x')!)
    let y = parseFloat(this.element.getAttribute('y')!)
    let width = parseFloat(this.element.getAttribute('width')!)
    let height = parseFloat(this.element.getAttribute('height')!)
    let isInXAxes = position.x <= x + width + (2* this.selectionOffset) + this.totalTranslation.x
                    && position.x >= x- this.selectionOffset + this.totalTranslation.x
    let isInYaxes = position.y <= y + height + (2* this.selectionOffset) + this.totalTranslation.y
                    && position.y >= y- this.selectionOffset + this.totalTranslation.y
    return isInXAxes && isInYaxes;
  }
  translate(translationPoint: Point): void {
    //throw new Error('Method not implemented.');
    this.totalTranslation.makeEqualTo(translationPoint);
    this.renderer.setAttribute(this.element, 'transform', `translate(${this.totalTranslation.x},${this.totalTranslation.y})`)
    this.sendProgressToServer(DrawingStatus.Selected);
  }
  scale(scalePoint: Point, direction: Point): void {
    //throw new Error('Method not implemented.');
    let x = parseFloat(this.element.getAttribute("x")!)
    let y = parseFloat(this.element.getAttribute("y")!)
    let width = parseFloat(this.element.getAttribute("width")!)
    let height = parseFloat(this.element.getAttribute("height")!)
    let minPoint = new Point(x , y)
    let maxPoint = new Point(x + width, y + height)
    if(direction.x === -1.0){
        //println(minPoint.x)
        minPoint.x += scalePoint.x
        //println(minPoint.x)
        this.currentX = minPoint.x
    }
    else if(direction.x === 1.0){
        maxPoint.x += scalePoint.x
    }
    if(direction.y === -1.0){
        minPoint.y += scalePoint.y
        this.currentY = minPoint.y
    }
    else if(direction.y === 1.0){
        maxPoint.y += scalePoint.y
    }

    if(minPoint.x >= maxPoint.x){
        direction.x *= -1
    }

    if(minPoint.y >= maxPoint.y){
        direction.y *= -1
    }
    this.renderer.setAttribute(this.element,"y", Math.min(minPoint.y,maxPoint.y).toString())
    this.renderer.setAttribute(this.element,"x", Math.min(minPoint.x,maxPoint.x).toString())
    this.renderer.setAttribute(this.element,"width", Math.abs(maxPoint.x - minPoint.x).toString())
    this.renderer.setAttribute(this.element,"height", Math.abs(maxPoint.y - minPoint.y).toString())
    this.sendProgressToServer(DrawingStatus.Selected)
  }
  getSelectionString(): void {
    //throw new Error('Method not implemented.');
    this.str += "<rect "
    let x = parseFloat(this.element.getAttribute("x")!)
    let y = parseFloat(this.element.getAttribute("y")!)
    let width = parseFloat(this.element.getAttribute("width")!)
    let height = parseFloat(this.element.getAttribute("height")!)
    let transform = this.element.getAttribute("transform")

    this.str += `x="${x}" `
    this.str += `y="${y}" `
    this.str += `width="${width}" `

    this.str += `height="${height}" `
    this.str += `transform="${transform}" `

    this.str += `stroke="#0000FF" `
    this.str += `stroke-width="3" `
    this.str += `fill="none" `
    this.str += `stroke-dasharray="4"`
    this.str += "/>\n"
  }
  calculateScalingPositions(): void {
    //throw new Error('Method not implemented.');
    this.scalingPositions.clear()
    let width = parseFloat(this.element.getAttribute("width")!)
    let height = parseFloat(this.element.getAttribute("height")!)
    let x = parseFloat(this.element.getAttribute("x")!)
    let y = parseFloat(this.element.getAttribute("y")!)
    let firstPos = new Point(x + this.totalTranslation.x, y + this.totalTranslation.y)
    let firstDirection = new Point(-1.0, -1.0)
    this.scalingPositions.set(firstPos,firstDirection)

    let secondPos = new Point(x + (width/2) + this.totalTranslation.x, y + this.totalTranslation.y)
    this.scalingPositions.set(secondPos, new Point(0.0,-1.0))

    let thirdPos = new Point(x + width + this.totalTranslation.x, y + this.totalTranslation.y)
    this.scalingPositions.set(thirdPos, new Point(1.0, -1.0))

    let forthPos = new Point(x + width + this.totalTranslation.x,
        y + (height/2) + this.totalTranslation.y)
    this.scalingPositions.set(forthPos, new Point(1.0, 0.0))

    let fifthPos = new Point(x + width + this.totalTranslation.x,
        y + height + this.totalTranslation.y)
    this.scalingPositions.set(fifthPos, new Point(1.0, 1.0))

    let sixthPos = new Point(x + (width/2) + this.totalTranslation.x
        , y + height + this.totalTranslation.y)
    this.scalingPositions.set(sixthPos, new Point(0.0, 1.0))

    let seventhPos = new Point(x + this.totalTranslation.x
        , y + height + this.totalTranslation.y)
    this.scalingPositions.set(seventhPos, new Point(-1.0, 1.0))

    let eighthPos = new Point(x + this.totalTranslation.x ,
        y + (height/2) + this.totalTranslation.y)
    this.scalingPositions.set(eighthPos, new Point(-1.0, 0.0))
  }
  getScalingPoint(point: Point): [Point, Point] | undefined {
    //throw new Error('Method not implemented.');
    for(let item of this.scalingPositions){
      let position = item[0]
      let x = position.x - RADUIS
      let y = position.y - RADUIS
      let width = (position.x + RADUIS) - x
      let height = (position.y + RADUIS) - y
      let inXAxes = point.x >= x && point.x <= x+ width
      let inYaxes = point.y >= y && point.y <= y+ height
      if(inXAxes && inYaxes){
        return item;
      }
    }
    return undefined;
  }
  getScalingPositionsString(): void {
    //throw new Error('Method not implemented.');
    let color: string = "";
    let mapEntries = userColorMap.entries();
    for(const entry of mapEntries){
      if(entry[1]!==undefined && entry[1] === this.userId){
        color = entry[0];
        break;
      }
    }
    this.calculateScalingPositions()
    for(let item of this.scalingPositions){
      console.log(item)
      let position = item[0];
      let x = position.x-RADUIS;
      let y = position.y- RADUIS;
      let width = (position.x + RADUIS) - x
      let height = (position.y + RADUIS) - y
      this.str += `<rect x=${x} y=${y} width=${width} height=${height} stroke=${color} fill=${color}></rect>\n`;
    }
  }

  parse(parceableString: string): void {
    //throw new Error('Method not implemented.');
    let matchX = X_REGEX.exec(parceableString);
    if(matchX=== null){
      console.log("problem with x regex");
    }
    else{
      this.renderer.setAttribute(this.element, "x", matchX[1]);
    }
    let matchY = Y_REGEX.exec(parceableString);
    if(matchY === null){
      console.log("problem with y regex");
    }
    else{
      this.renderer.setAttribute(this.element, 'y', matchY[1])
    }
    let matchWidth = WIDTH_REGEX.exec(parceableString);
    if(matchWidth === null){
      console.log('problem with width regex');
    }
    else{
      this.renderer.setAttribute(this.element,'width', matchWidth[1])
    }
    let matchHeight = HEIGHT_REGEX.exec(parceableString);
    if(matchHeight === null){
      console.log("problem with height regex");
    }
    else{
      this.renderer.setAttribute(this.element, 'height', matchHeight[1])
    }
    let matchTranslate = TRANSLATE_REGEX.exec(parceableString);
    if(matchTranslate === null){
      console.log("problem with with translate regex");
    }
    else{
      this.renderer.setAttribute(this.element, 'transform', `translate(${matchTranslate[1]},${matchTranslate[2]})`)
      this.totalTranslation.x = parseFloat(matchTranslate[1]);
      this.totalTranslation.y = parseFloat(matchTranslate[2])
    }
    let matchStroke = STROKE_REGEX.exec(parceableString);
    if(matchStroke === null){
      console.log("problem with stroke regex");
    }
    else{
      this.renderer.setAttribute(this.element, 'stroke', matchStroke[1])
    }
    let matchStrokeWidth = STROKE_WIDTH_REGEX.exec(parceableString);
    if(matchStrokeWidth === null){
      console.log("problem with stroke-width regex")
    }
    else{
      this.renderer.setAttribute(this.element, 'stroke-width', matchStrokeWidth[1])
    }
    let matchFill = FILL_REGEX.exec(parceableString);
    if(matchFill === null){
      console.log("problem with fill regex")
    }
    else{
      this.renderer.setAttribute(this.element, "fill", matchFill[1])
    }
    this.setCriticalValues();
  }
  unselect(): void {
    //throw new Error('Method not implemented.');
    //this.mouseIsDown = false;
    this.selected = false;
    if(this.contentId !== null && this.contentId !== undefined){
      this.sendProgressToServer(DrawingStatus.Done);
    }
  }
  delete(): void {
    //throw new Error('Method not implemented.');
    if(this.contentId !== null && this.contentId !== undefined){
      this.sendProgressToServer(DrawingStatus.Deleted);
    }
  }
  updateThickness(): void {
    let attr = this.interactionService.toolsAttributes$.value;
    this.attr = {shapeLineThickness: attr.shapeLineThickness, shapeType: attr.shapeType}
    if(this.attr.shapeType === ShapeTypes.OUTLINE){
      this.renderer.setAttribute(this.element, "fill", "none");
      this.renderer.setAttribute(this.element, "stroke", this.primaryColor);
    }
    else if(this.attr.shapeType === ShapeTypes.FULL){
      this.renderer.setAttribute(this.element, "stroke", "none");
      this.renderer.setAttribute(this.element, "fill", this.secondaryColor);
    }
    else{
      this.renderer.setAttribute(this.element, "stroke", this.primaryColor);
      this.renderer.setAttribute(this.element, "fill", this.secondaryColor);
    }
    this.renderer.setAttribute(this.element, "stroke-width", attr.shapeLineThickness!.toString())
    this.sendProgressToServer(DrawingStatus.Selected);
  }
  updatePrimaryColor(): void {
    //throw new Error('Method not implemented.');
    this.colorPick.colorSubject.subscribe((color: ChosenColors)=>{
      if(color!== undefined){
        this.primaryColor = color.primColor;
        if(this.attr.shapeType!== undefined){
          if(this.attr.shapeType === ShapeTypes.FULL){
            this.renderer.setAttribute(this.element, "stroke", DEFSEC)
          }
          else{
            this.renderer.setAttribute(this.element, "stroke", this.primaryColor)
          }
        }
      }
      else{
        this.primaryColor = DEFPRIM;
        this.renderer.setAttribute(this.element, "stroke", this.primaryColor)
      }
    })
    // To send To server
    this.sendProgressToServer(DrawingStatus.Selected);
  }
  updateSecondaryColor(): void {
    //throw new Error('Method not implemented.');
    this.colorPick.colorSubject.subscribe((color: ChosenColors)=>{
      if(color !== undefined){
        this.secondaryColor = color.secColor;
        if(this.attr.shapeType !== undefined){
          if(this.attr.shapeType === ShapeTypes.OUTLINE){
            this.renderer.setAttribute(this.element, "fill", DEFSEC)
          }
          else{
            this.renderer.setAttribute(this.element, "fill", this.secondaryColor)
          }
        }
      }
      else{
        this.secondaryColor = DEFSEC;
        this.renderer.setAttribute(this.element, "fill", this.secondaryColor)
      }
    })
    this.sendProgressToServer(DrawingStatus.Selected);
  }
  select(): void {
    //throw new Error('Method not implemented.');
    //this.mouseIsDown = true;
    this.selected = true;
    this.sendProgressToServer(DrawingStatus.Selected)
  }
  setCriticalValues(): void {
    //throw new Error('Method not implemented.');
    let xCert = parseFloat(this.element.getAttribute('x')!)
    let yCert = parseFloat(this.element.getAttribute('y')!)
    let widthCert = parseFloat(this.element.getAttribute('width')!)
    let heightCert = parseFloat(this.element.getAttribute('height')!)
    this.startTransformPoint.x = xCert + (widthCert/2)
    this.startTransformPoint.y = yCert + (heightCert/2)
  }
}
